src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs

```diff
@@ -55,7 +63,7 @@ public async Task Invoke(HttpContext context)
             if (waitInQueueTask.IsCompletedSuccessfully && !waitInQueueTask.Result)
             {
                 RequestThrottlingLog.RequestRejectedQueueFull(_logger);
-                context.Response.StatusCode = StatusCodes.Status503ServiceUnavailable;
+                await _requestThrottlingOptions.OnRejected(context);
```

Comment:
What if this throws? As it turns out, `_requestQueue.Release()` never gets called meaning the server will eventually stop accepting any new requests. This is ultimately the fault of not putting the start of the try/finally block in this method immediately after the call to `_requestQueue.TryEnterQueueAsync()`, but that's not the only thing that should be fixed.

Any time we call user-specified callbacks, we always wrap the call in a generic try/catch that swallows and logs any user-thrown exceptions. You can see an example of this where [Kestrel invokes the middleware pipeline.](https://github.com/aspnet/AspNetCore/blob/81f2d46660ec6ea92549edd55486de78a7597c58/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L628-L651)

Reply:
Kestrel's exception handling is special, it's at the top of the stack. A user callback throwing on the main request thread is fine, it gets logged by the server or exception handler and becomes a 500. No need to do anything extra so long as it the middleware state is maintained.

Reply:
In fact, swallowing the exception could lead to false 200 responses.

Reply:
So I don't need to change anything, right?

Reply:
With the first part of your comment @halter73, the code sure is super misleading. Also when implementing the other queuing strategy (that rejects at dequeue time) I needed to refactor this method in middleware to deal with the new flow, and then refactored it back after putting that on hold. Basically it's pretty fragile and I need to go back and work on it.

Reply:
> In fact, swallowing the exception could lead to false 200 responses.

That would need to be accounted for of course. That's one reason why I recommended adding a throwing OnRejected test which we should add anyway, but it might not make too much sense to do so until we add more functional testing to verify we get a 503/500 instead of a 200 response.

After in person discussion with @Tratcher, we decided that since this exception will still get thrown from the main middleware pipeline, and that should still get logged with a helpful StackTrace, we're fine not catching here.

I personally still think it's more correct to respond with a 503 instead of a 500 and swallow/log the exception, but I can understand not wanting to add the complexity.

> So I don't need to change anything, right?

No. Hopefully the above explains the reasoning.